### PR TITLE
util/log: implement dupFD per-platform

### LINKED ISSUE #14385: util/log: syscall.Dup2 and syscall.Stderr are not windows compatible
In clog.go:926 both syscall.Dup2 and syscall.Stderr are both used and they are both not compatible with windows.  
Please find alternatives as this is preventing a clean windows build.

### PR DESCRIPTION
**Pull request by tamird**
Fixes #14385.
Updates #14405.

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/14441)

### CODE REVIEW

File: pkg/util/log/stderr_redirect_windows.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import (
+    "syscall"
+)
+
+var (
+    kernel32         = syscall.MustLoadDLL("kernel32.dll")
+    procSetStdHandle = kernel32.MustFindProc("SetStdHandle")
+)
+
+func dupFD(oldfd uintptr, newfd uintptr) error {
```
Comment (petermattis) [inline pkg/util/log/stderr_redirect_windows.go]: Did you grab this from the internets somewhere? Can you provide a pointer?

**Review by petermattis — APPROVED** [review]
LGTM

**Comment by bdarnell** [comment]
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/util/log/stderr_redirect.go, line 42 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14441#-KgQGNcaSJG5DGByWZXp:-KgQGNcb-dDeaT1lLXw_:b-c5h46l) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f19305c87d0be870e60fef1fe57b4058c901c3d/pkg/util/log/stderr_redirect.go#L42)):*
> ```Go
> func restoreStderr() error {
> 	stderrRedirected = false
> 	return dupFD(OrigStderr.Fd(), os.Stderr.Fd())
> ```

I don't think this works. `OrigStderr.Fd()` is still 2, and file descriptor 2 was clobbered by `hijackStderr`. You need to `dup` (not `dup2`) stderr first in order to save a copy of it; if you don't it is lost for good. 

---

*[pkg/util/log/stderr_redirect_windows.go, line 27 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14441#-KgQFxrJXqvhNF6bSCxo:-KgQFxrJXqvhNF6bSCxp:b-xtdloe) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f19305c87d0be870e60fef1fe57b4058c901c3d/pkg/util/log/stderr_redirect_windows.go#L27)):*
> ```Go
> 
> func dupFD(oldfd uintptr, newfd uintptr) error {
>     r0, _, e1 := syscall.Syscall(procSetStdHandle.Addr(), 2, uintptr(oldfd), uintptr(newfd), 0)
> ```

These casts look unnecessary, `{old,new}fd` are already type `uintptr`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14441)*

**Comment by tamird** [comment]
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/util/log/stderr_redirect.go, line 42 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14441#-KgQGNcaSJG5DGByWZXp:-KgQU9_PrIbUP3QNlg1G:bri9e2g) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f19305c87d0be870e60fef1fe57b4058c901c3d/pkg/util/log/stderr_redirect.go#L42)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I don't think this works. `OrigStderr.Fd()` is still 2, and file descriptor 2 was clobbered by `hijackStderr`. You need to `dup` (not `dup2`) stderr first in order to save a copy of it; if you don't it is lost for good. 
</blockquote></details>

You're correct, fixed.

---

*[pkg/util/log/stderr_redirect_windows.go, line 26 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14441#-KgOqtQp36OqgJ8otHPT-r1-26:-KgQUBqFIhuBzOWuGj1f:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f19305c87d0be870e60fef1fe57b4058c901c3d/pkg/util/log/stderr_redirect_windows.go#L26)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Did you grab this from the internets somewhere? Can you provide a pointer?
</blockquote></details>

Done.

---

*[pkg/util/log/stderr_redirect_windows.go, line 27 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14441#-KgQFxrJXqvhNF6bSCxo:-KgQUCFDIi4TlowK46Mz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f19305c87d0be870e60fef1fe57b4058c901c3d/pkg/util/log/stderr_redirect_windows.go#L27)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

These casts look unnecessary, `{old,new}fd` are already type `uintptr`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14441)*

**Comment by petermattis** [comment]
:lgtm:

---

Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14441#-:-KgQcVe_IJfFPIwew2ta:bnfp4nl)*

**Comment by knz** [comment]
LGTM. Thank you much for this.

**Pull request merged**